### PR TITLE
Fix issue with rethrowing specific errors that were just caught

### LINKED ISSUE #10224: Compiler warning when throwing specific exception to outer general handler
### Summary of Problem
Occurs in #10221 too, but filed separately because it was unrelated.

C-compiler backend warning is generated when specific-exception handler in an inner exception try-catch block throws to a general-handler outer block.
```
In function: 'chpl_user_main':
warning: comparison of distinct pointer types lacks a cast
  throwingHandledError_chpl = (call_tmp_chpl3 == e_chpl);
                                              ^~
```
Code still executes:
```
Caught NilThrownError
Caught general error
```

### Steps to Reproduce

**Source Code:**

```chapel
module ExceptionWarning {
  proc main() {
    try {
      try {
        throw new NilThrownError();
      }
      catch e: NilThrownError {
        writeln("Caught NilThrownError");
        throw e;
      }
    }
    catch e {
      writeln("Caught general error");
    }
  }
} // module ExceptionWarning
```

**Compile command:**
No flags.

### Configuration Information
`chpl version 1.18.0 pre-release (e5cdbf130e)`

### PR DESCRIPTION
**Pull request by lydia-duncan**
We were getting a compilation warning/error about comparing distinct pointer
types without a cast when trying to rethrow an error that had been caught using
a specified subtype of Error.  E.g.

```chapel
try {
  ...
} catch e: SubError {
   throw e;
}
```

This change adds a cast when the type of the error being rethrown is not just
Error (and avoids adding a cast otherwise).

Resolves #10224 

Adds two tests to track this improved behavior (both of which failed before the
change).
- One where the rethrown error is in a top-level try/catch statement
- One where the rethrown error is in a nested try/catch statement (as provided
 in the issue which reported this problem)

Also adds five tests tracking classes that are not subtypes of Error:
- Throwing something that is not an error
- Catching something that is not an error
- Throwing and catching something that is not an error
- Catching something that is not an error and then rethrowing it
- Throwing, catching, and rethrowing something that is not an error

Passed a full paratest with futures

### CODE REVIEW

**Comment by lydia-duncan** [comment]
@mppf - I think you're probably best to review this?

File: compiler/passes/errorHandling.cpp [hunk 1]
```diff
@@ -420,9 +420,26 @@ bool ErrorHandlingVisitor::enterCallExpr(CallExpr* node) {
...
         VarSymbol* throwingHandledVar = newTemp("throwingHandledError", dtBool);
-        CallExpr*  throwingHandled    = new CallExpr(PRIM_EQUAL, thrownError, handledError);
-
         throwBlock->insertAtTail(new DefExpr(throwingHandledVar));
+
+        CallExpr* throwingHandled = NULL;
+
+        if (handledError->type != dtError) {
+          // We need to cast these to the same base type before comparing using
+          // pointer comparison if the type is a subtype of error
```
Comment (mppf) [inline compiler/passes/errorHandling.cpp]: Is other code ensuring that `handledError` is a subtype of `Error`? Maybe add a test of this?

File: test/errhandling/lydia/rethrownGenerally-nested.chpl [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Code from issue opened by BryantLam on Github
```
Comment (mppf) [inline test/errhandling/lydia/rethrownGenerally-nested.chpl]: I don't see any reason not to just put the issue # - e.g. code from issue #10224

File: test/errhandling/lydia/rethrownGenerally.chpl [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// Code from issue opened by BryantLam on Github
```
Comment (mppf) [inline test/errhandling/lydia/rethrownGenerally.chpl]: ditto

File: compiler/passes/errorHandling.cpp (hunk 1 above)
Comment (lydia-duncan) [reply compiler/passes/errorHandling.cpp]: It seems as though that is contingent on being able to throw something that is not a subtype of Error.  If you can't throw it, the catch block never triggers.  Trying to throw something that isn't a subtype of Error gives an error message, but not the most helpful one:
```
notAnErrorSubtype-thrown.chpl:5: In function 'main':
notAnErrorSubtype-thrown.chpl:7: error: unresolved call 'chpl_fix_thrown_error(unmanaged Foo)'
$CHPL_HOME/modules/internal/ChapelError.chpl:339: note: candidates are: chpl_fix_thrown_error(err: borrowed Error) [162960]
$CHPL_HOME/modules/internal/ChapelError.chpl:345: note:                 chpl_fix_thrown_error(type errType) [162981]
$CHPL_HOME/modules/internal/ChapelError.chpl:352: note:                 chpl_fix_thrown_error(err: unmanaged Error) [163059]
$CHPL_HOME/modules/internal/ChapelError.chpl:367: note:                 chpl_fix_thrown_error(ref err: owned Error) [163134]
```

Comment (lydia-duncan) [reply compiler/passes/errorHandling.cpp]: I wrote five variant tests, I'll commit them (some as futures) in this PR and open a separate issue for the better error message when throwing something that shouldn't be thrown

Comment (mppf) [reply compiler/passes/errorHandling.cpp]: Thanks!

**Pull request merged**